File: src/16_System_binarny/Zad10.py
"""
Otrzymujesz dwie liczby naturalne. Oblicz ile bitów 
należy odwrócić, aby zamienić liczbę A w liczbę B.  
"""


def bity_do_zmiany(liczba_a, liczba_b):
    '''
    Funkcja zwraca ilość bitów do zmiany liczby A w liczbę B.
    '''

    n = liczba_a ^ liczba_b

    licznik_jedynek = 0
    while n:
        n &= n - 1
        licznik_jedynek += 1

    return licznik_jedynek


def test_bity_do_zmiany():
    assert bity_do_zmiany(0, 0) == 0
    assert bity_do_zmiany(1, 2) == 31
    assert bity_do_zmiany(34, 73) == 5

if __name__ == "__main__":

    test_bity_do_zmiany()

File: src/cpp/16_system_binarny/zad10.cpp
/*
ZAD-10 — Ile bitów trzeba odwrócić (A → B)

**Poziom:** ★★☆
**Tagi:** `XOR`, `popcount`, `bitwise`

### Treść

Wczytaj dwie liczby naturalne `A` i `B`. Oblicz, ile bitów trzeba odwrócić w
`A`, aby otrzymać `B`.

### Wejście

* 1. linia: `A`
* 2. linia: `B`

### Wyjście

Jedna liczba naturalna: liczba różniących się bitów.

### Przykład

**Wejście:**

```
34
73
```

**Wyjście:**

```
5
```

*/
#include <cassert>

int bityDoZmiany(int liczbaA, int liczbaB) {
  /*
   * Funkcja zwraca ilosc bitow do zmiany liczby A w liczbe B.
   */
  int n = liczbaA ^ liczbaB;

  int licznik = 0;
  while (n != 0) {
    n &= (n - 1);
    licznik++;
  }

  return licznik;
}

void testBityDoZmiany() {
  int liczbaA = 34;
  int liczbaB = 73;
  int wynik = 5;

  assert(bityDoZmiany(liczbaA, liczbaB) == wynik);
}

int main() {
  testBityDoZmiany();

  return 0;
}
